### PR TITLE
Detection model of Mask-RCNN.

### PR DESCRIPTION
**Pull request by jerrywgz**

### CODE REVIEW

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py [hunk 1]
```diff
@@ -75,21 +111,29 @@ def build_input(self, image_shape):
                 name='gt_label', shape=[1], dtype='int32', lod_level=1)
             self.is_crowd = fluid.layers.data(
                 name='is_crowd',
-                shape=[-1],
+                shape=[1],
                 dtype='int32',
                 lod_level=1,
                 append_batch_size=False)
```
Comment (jerrywgz) [inline fluid/PaddleCV/faster_rcnn/models/model_builder.py]: is_crowd should be [M, 1]

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py [hunk 2]
```diff
@@ -237,15 +281,116 @@ def fast_rcnn_heads(self, roi_input):
...
-        #        label=labels_int64
-        #        )
-        cls_prob = fluid.layers.softmax(self.cls_score, use_cudnn=False)
-        loss_cls = fluid.layers.cross_entropy(cls_prob, labels_int64)
+        loss_cls = fluid.layers.softmax_with_cross_entropy(
+            logits=self.cls_score,
+            label=labels_int64,
+            numeric_stable_mode=True, )
+        #cls_prob = fluid.layers.softmax(self.cls_score, use_cudnn=False)
+        #loss_cls = fluid.layers.cross_entropy(cls_prob, labels_int64)
```
Comment (jerrywgz) [inline fluid/PaddleCV/faster_rcnn/models/model_builder.py]: can be removed

File: fluid/PaddleCV/faster_rcnn/README.md [hunk 3]
```diff
@@ -17,17 +17,20 @@ Running sample code in this directory requires PaddelPaddle Fluid v.1.0.0 and la
 
 ## Introduction
 
-[Faster Rcnn](https://arxiv.org/abs/1506.01497) is a typical two stage detector. The total framework of network can be divided into four parts, as shown below:
-<p align="center">
-<img src="image/Faster_RCNN.jpg" height=400 width=400 hspace='10'/> <br />
-Faster RCNN model
-</p>
+RCNN models are two stages detector. According to proposals and feature extraction, obtain class and more precise proposals.
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README.md]: Give the full name at the first time

File: fluid/PaddleCV/faster_rcnn/README.md [hunk 4]
```diff
@@ -17,17 +17,20 @@ Running sample code in this directory requires PaddelPaddle Fluid v.1.0.0 and la
...
 2. Region Proposal Network (RPN). RPN generates proposals for detection。This block generates anchors by a set of size and ratio and classifies anchors into fore-ground and back-ground by softmax. Then refine anchors to obtain more precise proposals using box regression.
 3. RoI Align. This layer takes feature maps and proposals as input. The proposals are mapped to feature maps and pooled to the same size. The output are sent to fc layers for classification and regression. RoIPool and RoIAlign are used separately to this layer and it can be set in roi\_func in config.py.
 4. Detection layer. Using the output of roi pooling to compute the class and locatoin of each proposal in two fc layers.
 
+[Mask Rcnn](https://arxiv.org/abs/1703.06870) is a classical instance segmentation model and an extension of Faster Rcnn
+
+Mask Rcnn is a two stage model as well. At the first stage, it generates proposals from input images; At the second stage, it obtains class result, bbox and mask which is the result from segmentation branch on original Faster Rcnn model. It decouples the relation between mask and classification.
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README.md]: ; At  -> . At

File: fluid/PaddleCV/faster_rcnn/README.md [hunk 5]
```diff
@@ -66,8 +69,10 @@ After data preparation, one can start the training step by:
        --model_save_dir=output/ \
        --pretrained_model=${path_to_pretrain_model}
        --data_dir=${path_to_data}
+       --MASK_ON=False
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README.md:72]: line 70, 71 后面需要加 \吧， 否则直接复制使用会有问题吧

再提供个run.sh的脚本吧？里面提供faster-rcnn， mask-rcnn的run脚本吧

File: fluid/PaddleCV/faster_rcnn/README_cn.md [hunk 6]
```diff
@@ -17,17 +17,21 @@
 
 ## 简介
 
-[Faster Rcnn](https://arxiv.org/abs/1506.01497) 是典型的两阶段目标检测器。如下图所示，整体网络可以分为4个主要内容：
-<p align="center">
-<img src="image/Faster_RCNN.jpg" height=400 width=400 hspace='10'/> <br />
-Faster RCNN 目标检测模型
-</p>
+RCNN系列模型为两阶段目标检测器。通过对图像生成候选区域，提取特征，判别特征类别并修正候选框位置。
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README_cn.md]: 同上，RCNN第一次出现的地方用全称

File: fluid/PaddleCV/faster_rcnn/README_cn.md [hunk 7]
```diff
@@ -17,17 +17,21 @@
...
+RCNN系列目前包含两个代表模型：Faster Rcnn，Mask Rcnn
+
+[Faster Rcnn](https://arxiv.org/abs/1506.01497) 整体网络可以分为4个主要内容：
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README_cn.md]: 应该是Rcnn，还是RCNN？

File: fluid/PaddleCV/faster_rcnn/config.py [hunk 8]
```diff
@@ -156,6 +162,15 @@
...
 _C.spatial_scale = 1. / 16.
 
+# resolution to represent mask labels
+_C.resolution = 14
+
+# Number of channels in the mask head
+_C.DIM_REDUCED = 256
+
+# Threshold for converting soft masks to hard masks
+_C.MRCNN_THRESH_BINARIZE = 0.5
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/config.py]: 一些变量控制是小写，一些是大写，请统一

File: fluid/PaddleCV/faster_rcnn/eval_helper.py [hunk 9]
```diff
@@ -106,38 +109,42 @@ def get_nmsed_box(rpn_rois, confs, locs, class_nums, im_info,
...
-        labels = im_results_n[:, -1]
+        boxes = im_results_n[:, 2:]
+        scores = im_results_n[:, 1]
+        labels = im_results_n[:, 0]
     im_results = np.vstack([im_results[k] for k in range(len(lod) - 1)])
     return new_lod, im_results
 
 
-def get_dt_res(batch_size, lod, nmsed_out, data):
+def get_dt_res(batch_size, lod, nmsed_out, data, numId_to_catId_map):
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/eval_helper.py]: numId_to_catId_map ->  num_id_to_cat_id_map ?

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py [hunk 10]
```diff
@@ -191,28 +230,29 @@ def rpn_heads(self, rpn_input):
...
-                input=roi_input,
-                rois=pool_rois,
-                pooled_height=cfg.roi_resolution,
-                pooled_width=cfg.roi_resolution,
-                spatial_scale=cfg.spatial_scale,
-                sampling_ratio=cfg.sampling_ratio)
-        rcnn_out = self.add_roi_box_head_func(pool)
-        self.cls_score = fluid.layers.fc(input=rcnn_out,
+        self.res5_2_sum = self.add_roi_box_head_func(roi_input, pool_rois)
+        self.rcnn_out = fluid.layers.pool2d(
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/models/model_builder.py]: 如果其他函数没用到，可以不用`self.xx=`吧，其他同

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py [hunk 11]
```diff
@@ -237,15 +277,112 @@ def fast_rcnn_heads(self, roi_input):
...
+            act=act_func,
+            param_attr=ParamAttr(
+                name='mask_fcn_logits_w', initializer=MSRA(uniform=False)),
+            bias_attr=ParamAttr(
+                name="mask_fcn_logits_b",
+                learning_rate=2.,
+                regularizer=L2Decay(0.)))
+        if self.is_train:
+            mask_fcn_logits = fluid.layers.lod_reset(mask_fcn_logits,
+                                                     self.mask_int32)
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/models/model_builder.py]: train分支不用lod_reset吧

File: fluid/PaddleCV/faster_rcnn/train.py [hunk 12]
```diff
@@ -92,8 +93,12 @@ def if_exist(var):
         fluid.io.load_vars(exe, cfg.pretrained_model, predicate=if_exist)
 
     if cfg.parallel:
+        exec_strategy = fluid.ExecutionStrategy()
+        #exec_strategy.num_threads = 64
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/train.py]: exec_strategy没用到

File: fluid/PaddleCV/faster_rcnn/utility.py [hunk 13]
```diff
@@ -85,6 +87,44 @@ def get_median_value(self):
         return np.median(self.deque)
 
 
+def print_as_dict(d):
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/utility.py]: 这个函数可以删了，下面没用到

File: fluid/PaddleCV/faster_rcnn/README.md (hunk 3 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README.md]: Done

File: fluid/PaddleCV/faster_rcnn/README.md (hunk 4 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README.md]: Done

File: fluid/PaddleCV/faster_rcnn/README.md (hunk 5 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README.md:72]: Added

File: fluid/PaddleCV/faster_rcnn/README_cn.md (hunk 6 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README_cn.md]: Done

File: fluid/PaddleCV/faster_rcnn/README_cn.md (hunk 7 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README_cn.md]: Done

File: fluid/PaddleCV/faster_rcnn/config.py (hunk 8 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/config.py]: Done

File: fluid/PaddleCV/faster_rcnn/eval_helper.py (hunk 9 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/eval_helper.py]: Done

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py (hunk 10 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/models/model_builder.py]: Done

File: fluid/PaddleCV/faster_rcnn/models/model_builder.py (hunk 11 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/models/model_builder.py]: Done

File: fluid/PaddleCV/faster_rcnn/train.py (hunk 12 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/train.py]: Done

File: fluid/PaddleCV/faster_rcnn/utility.py (hunk 13 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/utility.py]: Done

File: fluid/PaddleCV/faster_rcnn/README.md [hunk 14]
```diff
@@ -17,17 +17,20 @@ Running sample code in this directory requires PaddelPaddle Fluid v.1.0.0 and la
 
 ## Introduction
 
-[Faster Rcnn](https://arxiv.org/abs/1506.01497) is a typical two stage detector. The total framework of network can be divided into four parts, as shown below:
-<p align="center">
-<img src="image/Faster_RCNN.jpg" height=400 width=400 hspace='10'/> <br />
-Faster RCNN model
-</p>
+Region Convolutional Neural Network (RCNN) models are two stages detector. According to proposals and feature extraction, obtain class and more precise proposals.
+Now RCNN models contains two typical models: Faster RCNN and Mask RCNN.
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/README.md]: RCNN models contains ->  RCNN model contains

File: fluid/PaddleCV/faster_rcnn/train.py [hunk 15]
```diff
@@ -92,8 +93,11 @@ def if_exist(var):
         fluid.io.load_vars(exe, cfg.pretrained_model, predicate=if_exist)
 
     if cfg.parallel:
+        exec_strategy = fluid.ExecutionStrategy()
```
Comment (qingqing01) [inline fluid/PaddleCV/faster_rcnn/train.py]: 去掉 exec_strategy，这个没用到

File: fluid/PaddleCV/faster_rcnn/README.md (hunk 14 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/README.md]: Done

File: fluid/PaddleCV/faster_rcnn/train.py (hunk 15 above)
Comment (jerrywgz) [reply fluid/PaddleCV/faster_rcnn/train.py]: deleted

**Review by qingqing01 — APPROVED** [review]
